feat - Fixes #293. Shows tx fee estimate and total tx amount. See issues in commit details

ltfschoen opened the pull request:
* Remaining Issues:
  1. Update (28 Dec 2018): RESOLVED in commit https://github.com/paritytech/fether/pull/307/commits/13fd9f6e40d0e7fca0182e4c6ee114606d0a6050 - After entering a valid value and a recipient address, it works but the console displays error: `Uncaught TypeError: Cannot read property 'resolve' of null at flush (index.js:62)`

  2. Update (28 Dec 2018): REQUIRES this PR to be merged https://github.com/paritytech/fether/pull/326 - After I've enter a recipient and an amount and the Transaction Details have been calculated, then if the page refreshes (likely due to syncing) then the input values clear but the Transaction Details are still shown. Update: Note that I have raised this PR to prevent the input values from being cleared due to syncing https://github.com/paritytech/fether/pull/326 so that this will no longer be an issue

Screenshot:

<img width="355" alt="screen shot 2018-12-18 at 10 56 41 pm" src="https://user-images.githubusercontent.com/6226175/50180555-fcfd4300-031a-11e9-9569-ed0b8aa40ea8.png">

Tbaut commented:
Thanks a lot @ltfschoen  for the contributions. I'd like to keep the design as simple as possible, I'll try to come up with a Mockup soon to keep the screen simple because I find it overwhelming right now.

Tbaut commented:
Here is the most simple way to show what you've built, but only to interested users.
We will need this "details" dropdown if we want to let users chose the gas price (cf #306 )
![image](https://user-images.githubusercontent.com/33178835/50214872-ab3bd380-0381-11e9-8963-f6107aaa1f34.png)

ltfschoen commented:
@Tbaut I've pushed commits so it functions with show/hide details like in your wireframes.

I still have to resolve the two issues shown in the description.

ltfschoen commented on packages/fether-react/src/Send/TxForm/TxForm.js:
```diff
@@ -126,7 +176,35 @@ class Send extends Component {
+                          {estimatedTxFee &&
+                            !estimatedTxFee.isZero() && (
+                            <div>
+                              {showDetails && !isNaN(values.amount) ? (
```
Do you mean remove the 'Details' and 'Hide' buttons, and just show the details section all the time by default? Or retain the 'Details' and 'Hide' buttons functionality whilst preventing the "send" button from jumping when the user toggles one of the buttons?

ltfschoen replied:
I'll leave the 'Details' and 'Hide' button functionality, and just fix it so it doesn't jump each time you click 'Details' or 'Hide'

ltfschoen commented:
@Tbaut I've pushed more commits to address your latest review comments

I still have to resolve the three issues shown in the description.

ltfschoen commented:
@Tbaut I've pushed commits that fix the bugs. Suggest merging this PR https://github.com/paritytech/fether/pull/326 before this one so we double check that the second issue gets resolved by it

amaurymartiny commented on packages/fether-react/src/Send/TxForm/TxForm.js:
```diff
@@ -36,24 +37,50 @@ const MIN_GAS_PRICE = 3; // Safelow gas price from GasStation, in Gwei
+
   state = {
-    maxSelected: false
+    estimatedTxFee: ZERO,
```
Remove this from state. The estimated fee is final-form's `values.gas` time `values.gasPrice`, so no need for another state variable.

amaurymartiny commented on packages/fether-react/src/Send/TxForm/TxForm.js:
```diff
@@ -36,24 +37,50 @@ const MIN_GAS_PRICE = 3; // Safelow gas price from GasStation, in Gwei
+  componentWillUnmount () {
+    // Avoids encountering error `Can't call setState (or forceUpdate)
+    // on an unmounted component` when navigate from 'Send Ether' to 'Send THIBCoin'
+    this.isCancelled = true;
```
Is we remove `estimatedTxFee` (see above), we don't need isCancelled anymore.

amaurymartiny commented on packages/fether-react/src/Send/TxForm/TxForm.js:
```diff
@@ -98,6 +125,56 @@ class Send extends Component {
 
+    const { estimatedTxFee, showDetails } = this.state;
+
+    const renderFee = () => {
```
2 solutions here:
- either create a variable called `renderFee`
- or put this function as a class method.

Reason: now we're creating a new function on each render.

amaurymartiny commented on packages/fether-react/src/Send/TxForm/TxForm.js:
```diff
@@ -98,6 +125,56 @@ class Send extends Component {
+        .toString()} ETH (Gas Limit * Gas Price)`;
+    };
+
+    const renderCalculation = values => {
```
same as above.

amaurymartiny commented on packages/fether-react/src/Send/TxForm/TxForm.js:
```diff
@@ -98,6 +125,56 @@ class Send extends Component {
+      return `Calcs: (${gasLimitBn} WEI * ${gasPriceBn} GWEI)`;
+    };
+
+    const renderTotalAmount = values => {
```
same as above.

amaurymartiny commented on packages/fether-react/src/Send/TxForm/TxForm.js:
```diff
@@ -98,6 +125,56 @@ class Send extends Component {
+        .toString()} ETH`;
+    };
+
+    const renderDetails = values => {
```
same as above.

amaurymartiny commented on packages/fether-react/src/Send/TxForm/TxForm.js:
```diff
@@ -98,6 +125,56 @@ class Send extends Component {
+      )}`;
+    };
+
+    const showHideLabel = () => {
```
same as above.

amaurymartiny commented on packages/fether-react/src/Send/TxForm/TxForm.js:
```diff
@@ -98,6 +125,56 @@ class Send extends Component {
+      );
+    };
+
+    const showDetailsLabel = () => {
```
same as above.

ltfschoen commented:
@amaurymartiny I've pushed commits that address your latest review comments, such that the contents of the transaction details take your recommendation on board, `estimatedTxFee` has been removed from the state and `isCancelled` removed, the functions in the `render` function are now instance methods of a TxDetails component, and the text shown is now selectable.

See below screenshot:

<img width="360" alt="screen shot 2019-01-01 at 3 31 16 pm" src="https://user-images.githubusercontent.com/6226175/50570329-359d2200-0ddb-11e9-8990-49ca765a1f88.png">

ltfschoen commented:
@amaurymartiny I've created a commit https://github.com/paritytech/fether/commit/910f9c449a713f3198b700799840d36c512193b1 in a separate branch named 'luke-293-show-tx-fee-styled-components' where I've converted the code in this PR to use styled-components, as you suggested here in Issue #158 https://github.com/paritytech/fether/issues/158#issuecomment-450626560.

* Would you like to proceed using styled-components like this?

* Are you happy with how I've used ThemeProvider (see https://www.styled-components.com/docs/advanced)? Since I've only used ThemeProvider as an alternative to how we used Sass variables, perhaps I should wrap these  global styles around a parent component instead, such as fether-react/src/App/App.js similar to how it's done in this blogpost https://medium.com/styled-components/styled-components-getting-started-c9818acbcbbd?

* Where should we store the themes (i.e. `const theme = ...`)? Perhaps in assets/styledComponents/shared/_themes.js?
We may have a theme for fonts, and other things...

ltfschoen commented:
@amaurymartiny Update: I've been adding additional commits to integrate styled-components into Fether in branch https://github.com/paritytech/fether/compare/luke-293-show-tx-fee-styled-components

* `fetherTheme` theme is stored in paritytech/fether/packages/fether-react/src/assets/theme/globalStyle.js. Note that I have not used it yet though. See commit https://github.com/paritytech/fether/commit/5c5e04c43a7ecd32948cdaa663295a7499a58adb
* `ThemeProvider` is implemented as wrapper in App.js to pass the `fetherTheme` as prop `theme` to child components. See commit https://github.com/paritytech/fether/commit/5c5e04c43a7ecd32948cdaa663295a7499a58adb
* Shared styles that are used across multiple components (such as the previous CSS class `.window_content`) are now being stored in paritytech/fether/packages/fether-react/src/assets/theme/shared/styledComponents/DivWindowContent/DivWindowContent.js. See commit https://github.com/paritytech/fether/commit/5c5e04c43a7ecd32948cdaa663295a7499a58adb

What do you think of adding a feature to Fether somewhere that allows a user to choose different colour themes. This blogpost shows how easy it is to add a theme switcher https://medium.com/styled-components/styled-components-getting-started-c9818acbcbbd

amaurymartiny commented on packages/fether-react/src/Send/TxDetails/TxDetails.js:
```diff
@@ -0,0 +1,104 @@
+    const { estimatedTxFee } = this.props;
+
+    return `Fee: ${estimatedTxFee(values)
+      .div(10 ** 18)
```
Instead of div(10 ** 18), you can use the `fromWei` function from @parity/api/lib/util/wei

amaurymartiny commented on packages/fether-react/src/Send/TxDetails/TxDetails.js:
```diff
@@ -0,0 +1,104 @@
+
+    return `Total Amount: ${estimatedTxFee(values)
+      .plus(token.address === 'ETH' ? toWei(values.amount.toString()) : 0)
+      .div(10 ** 18)
```
`fromWei()`

amaurymartiny commented on packages/fether-react/src/Send/TxDetails/TxDetails.js:
```diff
@@ -0,0 +1,104 @@
+          {showDetails ? this.showHideAnchor() : this.showDetailsAnchor()}
+        </div>
+        <div className='form_field'>
+          <div hidden={!showDetails}>
```
TIL that div has a `hidden` prop. :)

amaurymartiny commented on packages/fether-react/src/Send/TxForm/TxForm.js:
```diff
@@ -39,21 +40,22 @@ class Send extends Component {
+
         if (this.preValidate(allValues) === true) {
-          return estimateGas(allValues, token, parityStore.api);
+          const newEstimatedTxFee = await estimateGas(
```
Hmm, I'm not sure what happens when `estimateGas` throws an error, it depends on how final-form-calculate is implemented. To be sure, let's add a try/catch here?

(^ not introduced by this PR, this potential bug was here before, I just saw it).

amaurymartiny commented on packages/fether-react/src/Send/TxForm/TxForm.js:
```diff
@@ -183,6 +203,14 @@ class Send extends Component {
                             type='range' // In Gwei
                           />
 
+                          <TxDetails
```
Here, we could do:
```
valid && isEstimatedTxFee(values) && !isNaN(values.amount) && <TxDetails ... />
```
so that:
- TxDetails is only mounted when necessary
- we can pass in less props to TxDetails (estimatedTxFee, token and values)

and remove this logic from the render function of TxDetails.

amaurymartiny commented on packages/fether-react/src/Send/TxDetails/TxDetails.js:
```diff
@@ -0,0 +1,104 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
```
Can you put TxDetails folder inside TxForm folder? Because that component is only used in TxForm

ltfschoen commented on packages/fether-react/src/Send/TxDetails/TxDetails.js:
```diff
@@ -0,0 +1,104 @@
+  };
+
+  renderDetails = values => {
+    return `${this.renderCalculation(values)}\n`
```
Unfortunately doing that makes it look like this:

<img width="314" alt="screen shot 2019-01-02 at 11 54 40 pm" src="https://user-images.githubusercontent.com/6226175/50592819-d417a900-0ee9-11e9-953f-8ab15d0ec783.png">

amaurymartiny replied:
The whitespaces are also taken into account, so you would need to move the `${this.renderFee(values)}` part to the beginning of line.

ltfschoen replied:
The alternatives I could get to work was having it all on one line like:
```
return `${this.renderCalculation(values)}\n${this.renderFee(values)}\n${this.renderTotalAmount(values)}`;
```

Or using `+` like:
```
    return `${this.renderCalculation(values)}\n` +
    `${this.renderFee(values)}\n` +
    `${this.renderTotalAmount(values)}`;
```

ltfschoen replied:
> The whitespaces are also taken into account, so you would need to move the `${this.renderFee(values)}` part to the beginning of line.

Yes that works and linting doesn't complain. But the code indentation looks like the following:

<img width="389" alt="screen shot 2019-01-03 at 1 31 47 pm" src="https://user-images.githubusercontent.com/6226175/50621782-5345c500-0f5c-11e9-892c-b356436eaaa1.png">

Another option that renders correctly is to add a couple of NUL chars at the beginning of those two lines so it doesn't look like the indentation needs to be fixed:

<img width="385" alt="screen shot 2019-01-03 at 1 52 31 pm" src="https://user-images.githubusercontent.com/6226175/50622087-d0723980-0f5e-11e9-821f-ab8d66220104.png">

ltfschoen commented:
@amaurymartiny I've pushed commits addressing your review comments

Tbaut commented:
- The "send" button is still jumping when the "details" appears.
- The color of the field when no selected is still too "light" to my eyes, it's good when selected though (click on the field)
![image](https://user-images.githubusercontent.com/33178835/50629906-d9eba980-0f3e-11e9-94e8-b986b255e4f2.png)
  selected:
![image](https://user-images.githubusercontent.com/33178835/50629903-d48e5f00-0f3e-11e9-8c5f-a8a3b90a227c.png)
- The "total amount" is truncated and therefore not exact (e.g when sending less than 0.0000000001) not sure this is a big problem though.

Tbaut commented:
Still happening on the latest iteration, actually, there is also a slight horizontal jump on top of the vertical one:
![image](https://i.imgur.com/Pi7K4lB.gif)

Tbaut commented on packages/fether-react/src/Send/TxForm/TxForm.js:
```diff
@@ -131,6 +159,16 @@ class Send extends Component {
+                          />
+
                           <Field
                             autoFocus
```
autofocus should be placed on the "to" field now.

ltfschoen commented:
> Still happening on the latest iteration, the send button jumps when "details" appears.
> There is also a slight horizontal jump of the whole windows (but this seems to be due to the automatic window size):
> ![image](https://camo.githubusercontent.com/7bf65cd82cff6ed6e5d65d7f1f5f0fec6b5bf622/68747470733a2f2f692e696d6775722e636f6d2f5069374b346c422e676966)

Oh yes I see it now on mine as well. I found that I just needed to hide the scroll on the body element

ltfschoen commented:
@Tbaut I've pushed commits addressing your issues. I noticed it on mine as well but hadn't looked close enough, and on closer inspection I found that it was the scroll bar appearing very briefly that caused the jumping effect. I found that I just needed to hide the scroll on the body element, which I've included in one of the commits

amaurymartiny commented on packages/fether-react/src/Send/TxForm/TxDetails/TxDetails.js:
```diff
@@ -0,0 +1,104 @@
+        ),
+        'ether'
+      )
+        .toFixed(18)
```
If you remove this line, then `.toString()` actually already removes trailing zeroes, and I believe `removeTrailingZeros` is not actually needed?

ltfschoen replied:
Yes removing that script and removing `.toFixed(18)` fixed it

ltfschoen commented:
Based on discussion with @Tbaut I've created a separate branch and this commit https://github.com/paritytech/fether/commit/bd56db628dfb68b15abc327ced11d8495180bfa0 that includes:
* Displays the "Details / Hide" buttons in the footer so when they appear they do not cause the height of the window to change (same horizontal alignment as the 'Send' / 'Checking...' button)
* When "Details" button is clicked it displays the transaction details on top of the other fields so it doesn't cause the height of the window to change.

With the code in that branch/commit the only issue that appears to still need fixing is when it changes to 'Checking...' instead of 'Send' it causes both the "Details / Hide" buttons and the transaction details popup to shift to the left (since the text of 'Checking...' has more characters and makes the button wider requiring more space).

If we'd like this option then it's probably worth also refactoring the code to try and use a SUI popup/modal. What do you think?

Screenshot:

<img width="360" alt="screen shot 2019-01-05 at 9 03 27 am" src="https://user-images.githubusercontent.com/6226175/50713603-bdc84380-10c9-11e9-82c5-10e0df82ff0c.png">

Tbaut commented:
I think there was a misunderstanding. My request was actually very small and most probably just needed a CSS trick. What you did before with a slide down was totally fine and matched the mockup we agreed on. 

The only problem I saw was with the jump of the "Send" button *when the details **button** appears*.
An image is worth 1000 words.
- We currently have a transition from 1 -> 2, the appearance of "Details" toggle button makes the windows longer, that's what I meant with the jump of the "Send" button.
- I wish to make sure the space taken by the "Details" button is already reserved (green area hilighted in 3) so that we would have a transition from 3 -> 2 (the green area in 3 should obviously be white).

![image](https://user-images.githubusercontent.com/33178835/50737325-09f9ad80-11c8-11e9-906b-460ecfedff58.png)

Please PM if that's not clear.
We could also simply show a disabled "Details" button when not available.

For the sake of completeness, here is my reasoning: 
- it's fine for things to change a lot (such as a window resize) when the user expects it, e.g user click on a "Details" button -> the windows shows a new section and resizes
- it's not fine if the user doesn't expect it, e.g the user enters an amount and an unexpected "Details" button appears which makes the window resize.

ltfschoen commented:
@Tbaut I've push a commit so there isn't a "green" area anymore by moving the "Details / Hide" toggle button div into the footer nav bar so it appears inline with the "Send / Checking..." button

Screenshots:

<img width="359" alt="screen shot 2019-01-07 at 5 44 06 am" src="https://user-images.githubusercontent.com/6226175/50740232-50d6ba00-123f-11e9-9972-a4c18f0478a8.png">

<img width="361" alt="screen shot 2019-01-07 at 5 44 18 am" src="https://user-images.githubusercontent.com/6226175/50740233-5207e700-123f-11e9-9ce8-ceadea1671fc.png">
